### PR TITLE
WIP sftp-store

### PR DESCRIPTION
**Pull request by mmb90**
Initially I tried to rebase your scp-store branch onto my master but there was lots of conflicts that were beyond my ability to resolve without really getting into the code. So I put that off and just copied the new blobstore. After tweaking with it for a few days I managed to get it to work. By work I mean my sftp_example works. I didn't write any unit tests for it. Here's where I've left it:

1) I don't know much about the few commits you have in the scp-store branch before WIP sftp store commit.
2) You can see what I've done in the code and via examples/sftp.
3) No unit tests but the example is a good starting point as a manual test.
4) I didn't do the fancy subpackage stuff for importing paramiko yet.
5) To test example/sftp_example.py I used the provenance-dev conda env and you can look at the example's readme.
6) I don't know if merging this pull request is the way to get the code in or not but it seemed like a good way to have the conversation.

Let me know if this is headed in the right direction.

### CODE REVIEW

**Comment by bmabey** [comment]
Overall looks good. I wouldn't worry about an automated test for this. I'll test it manually tomorrow and merge it in. Thanks!

BTW, with PR #23 you won't need to create the DB anymore in a startup script. :)

**Comment by bmabey** [comment]
Thanks! The subpackage is an easy change, just remove the `paramiko` out of requirements.txt and add an `extras_require` to the setup.py similar to done here:
https://github.com/dask/dask/blob/master/setup.py

Lets call this extra package `sftp`. So we should have:

```python
packages = ['provenance', 'provenance.sftp']
```

Once that is done I'll merge this in and can cut a release if you would like.

**Comment by mmb90** [comment]
Now do my examples need `import provenance.sftp` in addition to `import provenance as p` so the SFTPStore class exists when I try to use it?

**Comment by mmb90** [comment]
The build failed and I'm guessing the code for the SFTPStore needs to be moved out of provenance/blobstore.py and into provenance/sftp/sftp_blobstore.py?

**Comment by bmabey** [comment]
Yeah, based on the error message it looks like it is looking for a `provenance/sftp` dir. I would just put it in `provenance/sftp/_init_.py`. To test this locally run `make dist`

**Comment by bmabey** [comment]
This looks good and can merge in as is. The only other change that may be worth doing it to move the sftp blobstore code into the `sftp/__init__.py` and then update the `_config` module to try to import it and add it to the `BLOBSTORE_TYPES` dict.

**Comment by mmb90** [comment]
I'll give that a try now.

**Comment by mmb90** [comment]
Question. The example won't run because `paramiko` is not found when `provenance.sftp` is imported in the `_config` module. Is it supposed to be installed automatically b/c of the `extras_require` or should there be a `try` around `import paramiko` that can notify the user to install it?

File: provenance/_config.py [hunk 1]
```diff
@@ -5,6 +5,11 @@
 import copy
 import toolz as t
 
+try:
+    import provenance.sftp as sftp
+except ImportError:
```
Comment (bmabey) [inline provenance/_config.py]: I would move this down to when you try to add it to the dict... so:

```python
try:
    import provenance.sftp as sftp
    BLOBSTORE_TYPES['sftp'] = sftp.SFTPStore
except ImportError:
    pass
```

I wouldn't worry about printing a message for now. Ideally, an exception would be thrown when they tried to use the SFTP store.. which I guess you could do something like:

```python
try:
    import provenance.sftp as sftp
    BLOBSTORE_TYPES['sftp'] = sftp.SFTPStore
except ImportError as e:

    class SFTPStore(object):
        def __init__(*args, **kargs):
            raise(e)

    BLOBSTORE_TYPES['sftp'] = SFTPStore
```

**Comment by bmabey** [comment]
The way subpackages work with `pip` is that you have to request that the extra ones be installed like so: `pip install provenance[sftp]` or you can get all of them with `pip install provenance[all]`.

Unfortunately I don't think `conda` doesn't support subpackages when installing the package from github like this. You may be able to use the brackets once a package has been published. In any case the way to workaournd it is to add `paramiko` to your env file. This isn't ideal I realize but that is the only option.

**Comment by mmb90** [comment]
@bmabey ok! I think we are there now. Everything is working as far as I can tell. Sorry for my subpackaging ignorance. I tried to read up on it but it seems to be not very well documented. (Maybe it is if you already know what your looking for.) To answer your earlier question, yes, I would like a new version after this merge, thanks.

**Comment by bmabey** [comment]
@mmb90 thanks for all the work on this!

I've just released [version 0.9.4](https://github.com/bmabey/provenance/commit/662a04437180b7588ca8f97d157127d2ab30d57a).

**Pull request merged**